SMLHelper/BiomeThings/FloatingOrigin.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Based on the Unity Wiki FloatingOrigin script by Peter Stirling
+// URL: http://wiki.unity3d.com/index.php/Floating_Origin
+
+using UnityEngine;
+using UnityEngine.SceneManagement;
+
+internal class FloatingOrigin : MonoBehaviour
+{
+    internal static Vector3 CurrentOffset;
+    [Tooltip("Point of reference from which to check the distance to origin.")]
+    public Transform ReferenceObject;
+
+    [Tooltip("Distance from the origin the reference object must be in order to trigger an origin shift.")]
+    public float Threshold = 500f;
+
+    [Header("Options")]
+    [Tooltip("When true, origin shifts are considered only from the horizontal distance to orign.")]
+    public bool Use2DDistance = true;
+
+    [Tooltip("When true, updates ALL open scenes. When false, updates only the active scene.")]
+    public bool UpdateAllScenes = true;
+
+    [Tooltip("Should ParticleSystems be moved with an origin shift.")]
+    public bool UpdateParticles = true;
+
+    [Tooltip("Should TrailRenderers be moved with an origin shift.")]
+    public bool UpdateTrailRenderers = true;
+
+    [Tooltip("Should LineRenderers be moved with an origin shift.")]
+    public bool UpdateLineRenderers = true;
+
+    private ParticleSystem.Particle[] parts = null;
+
+    void LateUpdate()
+    {
+        if (ReferenceObject == null)
+            return;
+        if (!LargeWorldStreamer.main.IsWorldSettled())
+            return;
+        Vector3 referencePosition = ReferenceObject.position;
+
+        if (Use2DDistance)
+            referencePosition.y = 0f;
+
+        if (referencePosition.magnitude > Threshold)
+        {
+            CurrentOffset = referencePosition;
+            MoveRootTransforms(referencePosition);
+
+            if (UpdateParticles)
+                MoveParticles(referencePosition);
+
+            if (UpdateTrailRenderers)
+                MoveTrailRenderers(referencePosition);
+
+            if (UpdateLineRenderers)
+                MoveLineRenderers(referencePosition);
+        }
+    }
+
+    private void MoveRootTransforms(Vector3 offset)
+    {
+        if (UpdateAllScenes)
+        {
+            for (int z = 0; z < SceneManager.sceneCount; z++)
+            {
+                foreach (GameObject g in SceneManager.GetSceneAt(z).GetRootGameObjects())
+                {
+                    if(g.name == "MainCamera (UI)" || g.name == "MainCamera")
+                    {
+                        continue;
+                    }else if(g.TryGetComponent<uGUI_BuilderMenu>(out var buildermenu))
```

Comment:
Please run a the automatic code formatting on all updated and new files.

This will minimize unintended whitespace changes going forward.